index.js

```diff
@@ -1,23 +1,115 @@
 var FlumeReduce = require('flumeview-reduce')
+var pull = require('pull-stream')
 var ref = require('ssb-ref')
 var get = require('lodash/get')
 var merge = require('lodash/merge')
+var map = require('lodash/map')
+var sortBy = require('lodash/sortBy')
+var entries = require('lodash/entries')
+var has = require('lodash/has')
 var TagHelper = require('scuttle-tag')
 
 exports.name = 'tags'
 exports.version = require('./package.json').version
 exports.manifest = {
   stream: 'source',
-  get: 'async'
+  get: 'async',
+  streamTaggedMessages: 'source',
+  getTaggedMessages: 'async',
+  streamMessagesTaggedWith: 'source',
+  getMessagesTaggedWith: 'async'
 }
 
 var initialState = {}
 
 exports.init = function (ssb, config) {
   var ScuttleTag = TagHelper(ssb)
-  return ssb._flumeUse('tags', FlumeReduce(2, reduce, map, null, initialState))
+  var view = ssb._flumeUse('tags', FlumeReduce(3, tagsReduce, tagsMap, null, initialState))
 
-  function reduce (result, item) {
+  return {
+    stream: view.stream,
+    get: view.get,
+    streamTaggedMessages,
+    getTaggedMessages,
+    streamMessagesTaggedWith,
+    getMessagesTaggedWith
+  }
+
+  function getTaggedMessages (cb) {
+    view.get((err, data) => {
+      if (err) cb(err)
+      var messages = {}
+      for (var author in data) {
+        for (var tag in data[author]) {
+          for (var msg in data[author][tag]) {
+            var timestamp = data[author][tag][msg]
+            if (!messages[msg] || messages[msg] < timestamp) messages[msg] = timestamp
+          }
+        }
+      }
+      var sortedMsgIds = map(sortBy(entries(messages), ([msgId, timestamp]) => -timestamp), 0)
+      cb(null, sortedMsgIds)
+    })
+  }
+
+  function streamTaggedMessages (opts) {
+    return pull(
+      view.stream(opts),
+      pull.map((item) => {
+        if (item.tagKey) return item
+        var messages = {}
+        for (var author in item) {
+          for (var tag in item[author]) {
+            for (var msg in item[author][tag]) {
+              var timestamp = item[author][tag][msg]
+              if (!messages[msg] || messages[msg] < timestamp) messages[msg] = timestamp
+            }
+          }
+        }
+        return map(sortBy(entries(messages), ([msgId, timestamp]) => -timestamp), 0)
+      })
+    )
+  }
+
+  function getMessagesTaggedWith (tagId, cb) {
+    view.get((err, data) => {
+      if (err) cb(err)
+      var messages = {}
+      for (var author in data) {
+        for (var msg in data[author][tagId]) {
+          var timestamp = data[author][tagId][msg]
+          if (!messages[msg] || messages[msg] < timestamp) messages[msg] = timestamp
+        }
+      }
+      var sortedMsgIds = map(sortBy(entries(messages), ([msgId, timestamp]) => -timestamp), 0)
+      cb(null, sortedMsgIds)
+    })
+  }
```

Comment:
This is the case I wrote the example code for on scuttlebutt using ssb-query.
Note that my code did not handle the reducing down to check for things being untagged.